### PR TITLE
Add `webpki-roots` and `native-certs` crate features, take 2

### PR DESCRIPTION
**Pull request by daxpedda**
This splits the changes from #1943 into two parts. This is the first part, only introducing loading the native certificates. The changes are basically only touching the code that was loading certificates from `webpki-roots` and introducing the new crate features.

I could split this up further into `trust-dns-proto` and `trust-dns-resolver`?

This implementation has the big downside that `trust-dns-resolver` will attempt to load the native certificates only once and save the `Result` in a `Lazy`. So there currently no way for the user to re-attempt or reload the native certificates. This will be addressed in the second part.

Replaces #1943.

### CODE REVIEW

File: crates/resolver/src/quic.rs
Comment (daxpedda) [inline crates/resolver/src/quic.rs:34]: I could remove the returned `Result` here by building a `QuicClientConnect` `Future` that returns the error instead.

Comment (bluejekyll) [reply crates/resolver/src/quic.rs:34]: The reason I ended up with a lot of those Connect types and the others, was to ensure that we had a consistent behavior to deal with in terms of Futures. I think it makes for a simpler API to return something like that Connect type. If others have a differing opinion I'm open to that.

File: crates/proto/Cargo.toml
Comment (djc) [inline crates/proto/Cargo.toml:43]: As written, this doesn't make much sense to me. Do you intend to expand the list of features that `webpki-roots`/`native-certs` requires in future PRs?

**Review by djc — COMMENTED** [review]
Sorry, being nitpicky about the commit history, because I know there's a bunch more large changes here and I'd like to make reviewing them easy.

I'm inclined to say the `Lazy<Arc<ClientConfig>>` design doesn't make all that much sense and we should instead use a design where a "global" default config lives in something like the `ResolverConfig`, `NameServerConfigGroup` or `ResolverOpts` but maybe this should be left for a future PR?

File: crates/proto/Cargo.toml
Comment (daxpedda) [reply crates/proto/Cargo.toml:43]: > As written, this doesn't make much sense to me.

Sorry, not sure where you are coming from here.

I didn't make it require anything else on purpose because it's a shared feature between `dns-over-tls`, `dns-over-quic` and so on.

How about renaming it to `dns-over-rustls-webpki-root`?
Or did you specifically mean we should just require `dns-over-rustls` here?

File: crates/proto/src/https/https_client_stream.rs
Comment (daxpedda) [reply crates/proto/src/https/https_client_stream.rs:737]: This PR specifically only touches the loaded certificates, I didn't separate the configurations, I just left everything else as is. I don't know what the original motivation to name this function `tls12` was, I would argue that it's actually wrong. Again, this PR didn't make or name this function.

File: crates/proto/src/quic/quic_client_stream.rs
Comment (daxpedda) [reply crates/proto/src/quic/quic_client_stream.rs:304]: This isn't unrelated. The original import of `rustls::OwnedTrustAnchor` doesn't make sense anymore if `webpki-roots` isn't active, so this is required unless we want to use `#[allow(unused_import)]` (or w/e it's called).

File: crates/resolver/examples/custom_provider.rs
Comment (daxpedda) [reply crates/resolver/examples/custom_provider.rs:3]: I was following the other examples here, but for sure `required-features` is the proper solution here.

File: crates/resolver/src/https.rs
Comment (daxpedda) [reply crates/resolver/src/https.rs:36]: This isn't a stylistic change, I need to return an error here if appropriate. I'm not aware of an alternative to `map_or_else` that supports returning an error.

File: crates/proto/src/https/https_client_stream.rs
Comment (daxpedda) [reply crates/proto/src/https/https_client_stream.rs:737]: This function is behind a `#[cfg(test)]` guard btw and is only used in these tests, maybe that caused the confusion?

File: crates/proto/Cargo.toml
Comment (daxpedda) [reply crates/proto/Cargo.toml:34]: This was actually an unused dependency that I removed, I separated that removal into it's own commit now.

**Comment by daxpedda** [comment]
> Sorry, being nitpicky about the commit history, because I know there's a bunch more large changes here and I'd like to make reviewing them easy.

This is totally fine! I'm happy to do the legwork here and make it as easy as possible to review!

> I'm inclined to say the `Lazy<Arc<ClientConfig>>` design doesn't make all that much sense and we should instead use a design where a "global" default `Arc<ClientConfig>` (or maybe even only the `RootCertStore`?) lives in something like the `ResolverConfig`, `NameServerConfigGroup` or `ResolverOpts` but maybe this should be left for a future PR? (Maybe this is already what you were planning on doing in a future PR.)

I think the big upside of `Lazy<Arc<ClientConfig>>` is for the bundled certificates to not consume memory more then once, but that doesn't play well with the native certs as we've seen in #1943.

Unfortunately storing a `RootCertStore` doesn't really work because we would have to clone it for every `ClientConfig` created from it. We could store a sharable [`WebPkiVerifier`](https://docs.rs/rustls/0.21.5/rustls/client/struct.WebPkiVerifier.html) but that's not possible without enabling `rustls/dangerous_configuration`, I've described the issue in more detail here: https://github.com/bluejekyll/trust-dns/issues/1990#issue-1816367149. I've worked around that in #2001 by creating a base `ClientConfig` that can be cloned, which just clones a bunch of `Arc`s, to reuse the root certificates without consuming additional memory.

As for where to store it I would argue that a "global" default would make sense to live in `ResolverOpts`.

Personally I'm also in favor of moving away from the `Lazy` and my preference would be to store it instead in `ResolverOpts`.

Let me know what you think!

File: crates/resolver/src/quic.rs
Comment (daxpedda) [reply crates/resolver/src/quic.rs]: Done.

File: crates/resolver/examples/custom_provider.rs
Comment (daxpedda) [reply crates/resolver/examples/custom_provider.rs:3]: Ah ... `required-features` doesn't work here because it needs an OR ..., I will put at least `tokio-runtime` into the `required-features`.

Comment (daxpedda) [reply crates/resolver/examples/custom_provider.rs]: Alright, adjusted the other examples as well.

File: crates/proto/src/quic/quic_client_stream.rs
Comment (djc) [reply crates/proto/src/quic/quic_client_stream.rs:304]: Ah, fair.

File: crates/proto/src/https/https_client_stream.rs
Comment (djc) [reply crates/proto/src/https/https_client_stream.rs:737]: I'm aware this is the pre-existing situation. I think we should assume there wasn't a great reason (or perhaps the reason was that rustls originally required separate configs for QUIC) to have a separate 1.2/1.3 configs and so it would be nice to have a prefix commit that deduplicates this.

File: crates/proto/Cargo.toml
Comment (djc) [reply crates/proto/Cargo.toml:43]: Sorry, I meant it's unnecessary to redirect the optional webpki-roots dependency through a `webpki-roots` feature.

**Comment by djc** [comment]
> I've worked around that in #2001 by creating a base `ClientConfig` that can be cloned, which just clones a bunch of `Arc`s, to reuse the root certificates without consuming additional memory.
> 
> As for where to store it I would argue that a "global" default would make sense to live in `ResolverOpts`.
> 
> Personally I'm also in favor of moving away from the `Lazy` and my preference would be to store it instead in `ResolverOpts`.

This sounds like a good direction!

**Comment by daxpedda** [comment]
CI currently fails on a test called `test_sec_lookup_fails`, doesn't work for me locally on `main` either so I believe this is unrelated to this PR. Can be reproduced with:
`cargo test -p trust-dns-resolver --all-features -- test_sec_lookup_fails`

File: crates/resolver/src/https.rs
Comment (daxpedda) [reply crates/resolver/src/https.rs]: This is now reworked in https://github.com/bluejekyll/trust-dns/pull/2005/commits/3a21b5aee0ea4548f8bf33b6a2de08ca37e4a7f0 anyways.

File: crates/proto/Cargo.toml
Comment (daxpedda) [reply crates/proto/Cargo.toml]: Ah sorry, yeah, removed it.

File: crates/proto/src/https/https_client_stream.rs
Comment (daxpedda) [reply crates/proto/src/https/https_client_stream.rs:737]: Apologies, I'm not reading very carefully today, you said that from the start, I was assuming too much here :sweat_smile:!

I was just about to do that when I stumbled on the fact that unlike DoQ, DoT and DoH don't actually have a default `ClientConfig` in `trust-dns-proto`, so that's why this function is here. See [`trust_dns_proto::quic::client_config_tls13_webpki_roots`](https://docs.rs/trust-dns-proto/latest/trust_dns_proto/quic/fn.client_config_tls13_webpki_roots.html).

I could add an equivalent for DoT/DoH if you would like me to?

Comment (djc) [reply crates/proto/src/https/https_client_stream.rs:737]: So DoT and DoH get their config from the `NameServerConfigGroup`? I guess it might be better to live with the duplication in this PR and then get the configs straightened out in your next one?

Comment (daxpedda) [reply crates/proto/src/https/https_client_stream.rs:737]: Sounds good to me!

**Comment by daxpedda** [comment]
> CI currently fails on a test called `test_sec_lookup_fails`, doesn't work for me locally on `main` either so I believe this is unrelated to this PR. Can be reproduced with: `cargo test -p trust-dns-resolver --all-features -- test_sec_lookup_fails`

I guess https://trust-dns.org is down?

**Review by djc — APPROVED** [review]

**Comment by daxpedda** [comment]
> > CI currently fails on a test called `test_sec_lookup_fails`, doesn't work for me locally on `main` either so I believe this is unrelated to this PR. Can be reproduced with: `cargo test -p trust-dns-resolver --all-features -- test_sec_lookup_fails`
> 
> I guess https://trust-dns.org is down?

I temporarily disabled the test to make the CI run. Let me know if you want me to remove it before merging.

**Comment by djc** [comment]
I'm going to leave it to @bluejekyll to decide if he wants to merge this with the ignored test (or fix up his DNS/web server setup?).

**Comment by daxpedda** [comment]
If we decide that we don't want to merge this now because it's a breaking change, let me know. There are other PRs I want to work on that (hopefully) don't require a breaking change, but I am waiting on this because I don't want to spend time constantly rebasing everything.

No stress though, I'm not in a hurry!

**Comment by djc** [comment]
I think we pretty much default to only doing breaking releases except if there's a security response needed.

**Comment by daxpedda** [comment]
I removed the commit disabling the test.

The last commit "Fix CI" was probably necessary because of the 1.72 release and has nothing to do with this PR. I'm happy to split it into a separate PR if needed.

**Comment by djc** [comment]
Cool, look forward to merging this. 

> The last commit "Fix CI" was probably necessary because of the 1.72 release and has nothing to do with this PR. I'm happy to split it into a separate PR if needed.

Separate PR is unnecessary from my side, but would like one commit for `cargo fmt` and one commit for clippy fixes.

**Comment by bluejekyll** [comment]
Sorry for the delay on this. I'm reviewing this now.

File: crates/proto/src/https/https_client_stream.rs
Comment (bluejekyll) [inline crates/proto/src/https/https_client_stream.rs:757]: we should change the result type of this function to `Result<ClientConfig, ...>` and return an error here. We don't want to panic unless absolutely necessary.

**Review by bluejekyll — CHANGES_REQUESTED** [review]
Thank you for all the cleanup on the examples and configuration as well. This looks great. You clearly spent a lot of time and energy on this change, I really appreciate all that effort.

See the one comment about the panic, once that is fixed, I will happily merge this in.

File: crates/proto/src/https/https_client_stream.rs
Comment (djc) [reply crates/proto/src/https/https_client_stream.rs:757]: IIRC this is for tests only, so panicking is fine.

Comment (bluejekyll) [reply crates/proto/src/https/https_client_stream.rs:757]: oh, you're right. It's in the test module, I should have expanded the review more. Thanks for pointing that out, @djc

**Review by bluejekyll — APPROVED** [review]
ok, this is an awesome change. Thank you for all the work!

**Comment by daxpedda** [comment]
> Separate PR is unnecessary from my side, but would like one commit for `cargo fmt` and one commit for clippy fixes.

Apologies, never got to separate those commits!

Thank you all for the guidance, was a great experience!

**Pull request merged**
